### PR TITLE
CORDA-3220: Updating FinalityFlow with functionality to specify StatesToRecord

### PR DESCRIPTION
**Pull request by rogersanick**
👮🏻👮🏻👮🏻 !!!! DESCRIBE YOUR CHANGES HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

Updating FinalityFlow with functionality to allow the initiator to indicate the appropriate StatesToRecord. This allows the initiating party to record states in transactions which they are proposing, but are not necessarily participants of. 

An example would be the issuance of a token where the only participant is the holder NOT the issuer.

JIRA Tickets - Corda-3220

# PR Checklist:

- [✅ ] Have you run the unit, integration and smoke tests as described [here](https://docs.corda.net/head/testing.html)?
- [✅ ] If you added public APIs, did you write the JavaDocs?
- [✅ ] If the changes are of interest to application developers, have you added them to the [changelog](https://github.com/corda/corda/blob/master/docs/source/changelog.rst), and potentially the [release notes](https://github.com/corda/corda/blob/master/docs/source/release-notes.rst)?
- [✅ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/corda/corda/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/corda/corda/blob/master/CONTRIBUTING.md#developer-certificate-of-origin).

Thanks for your code, it's appreciated! :)

### CODE REVIEW

**Comment by rogersanick** [comment]
- Closed old PR and reopened this one with a new branch name (nickRogersFinalityFlowFix). Branches prefaced with 'release' have specific admin permissions applied to them.
- Adjusted target branch to 4.4 (new branch cut)
- Added in feedback regarding kDoc updates.

File: core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt [hunk 1]
```diff
@@ -68,13 +69,15 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
...
+     * @param statesToRecord Which transactions to commit to the vault. Defaults to [StatesToRecord.ONLY_RELEVANT].
      */
     @JvmOverloads
     constructor(
             transaction: SignedTransaction,
             sessions: Collection<FlowSession>,
-            progressTracker: ProgressTracker = tracker()
-    ) : this(transaction, emptyList(), progressTracker, sessions, true)
+            progressTracker: ProgressTracker = tracker(),
+            statesToRecord: StatesToRecord = ONLY_RELEVANT
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt]: The API check is failing:

```
[14:56:45]	[Step 7/14] Number of API removals/changes: 1
[14:56:45]	[Step 7/14] -  public <init>(net.corda.core.transactions.SignedTransaction, java.util.Collection, net.corda.core.utilities.ProgressTracker, int, kotlin.jvm.internal.DefaultConstructorMarker)
```

You wouldn't know what the issue is by looking at the error message. The c'tor that is missing is a "hidden" one that's needed by apps written in Kotlin. Whenever a method or c'tor has default values then a special overload is created with `kotlin.jvm.internal.DefaultConstructorMarker` at then end. It's an implementation detail of Kotiln's default values feature.

However the downside is that by adding a new parameter here we've deleted this c'tor (and created a new one which is exactly the same but with the new parameter).

TL;DR - What you need to do is leave this c'tor as is and create a new one with `statesToRecord`.

File: core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt [hunk 2]
```diff
@@ -68,6 +69,7 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
      * @param transaction What to commit.
      * @param sessions A collection of [FlowSession]s for each non-local participant of the transaction. Sessions to non-participants can
      * also be provided.
+     * @param statesToRecord Which transactions to commit to the vault. Defaults to [StatesToRecord.ONLY_RELEVANT].
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt]: Delete

File: core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt [hunk 3]
```diff
@@ -76,6 +78,21 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
...
+     *
+     * @param transaction What to commit.
+     * @param sessions A collection of [FlowSession]s for each non-local participant of the transaction. Sessions to non-participants can
+     * also be provided.
+     * @param statesToRecord Which transactions to commit to the vault. Defaults to [StatesToRecord.ONLY_RELEVANT].
+     */
+    constructor(
+            transaction: SignedTransaction,
+            sessions: Collection<FlowSession>,
+            progressTracker: ProgressTracker = tracker(),
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt]: Move this param to the end

File: core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt (hunk 1 above)
Comment (rogersanick) [reply core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt]: Thanks @shamsasari! I just added the new constructor. Hopefully, this resolves the issue. 

I assumed that the @ JVMOverloads annotation would solve this, but it seems like it just generates conditionals that call with different default values in different circumstances.

File: core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt [hunk 4]
```diff
@@ -76,6 +78,21 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
...
 
+    /**
+     * Notarise the given transaction and broadcast it to all the participants.
+     *
+     * @param transaction What to commit.
+     * @param sessions A collection of [FlowSession]s for each non-local participant of the transaction. Sessions to non-participants can
+     * also be provided.
+     * @param statesToRecord Which transactions to commit to the vault. Defaults to [StatesToRecord.ONLY_RELEVANT].
+     */
+    constructor(
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:93]: You will need to add `@JvmOverloads` because of the default tracker value.

Comment (rogersanick) [reply core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:93]: Added!

File: core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt (hunk 3 above)
Comment (rogersanick) [reply core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt]: Assuming you're referring to the progressTracker param? So the order would be:

transaction, sessions, statesToRecord, progressTracker

File: core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt (hunk 2 above)
Comment (rogersanick) [reply core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt]: Deleted! Thanks.

File: core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt [hunk 5]
```diff
@@ -76,6 +77,22 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
             progressTracker: ProgressTracker = tracker()
     ) : this(transaction, emptyList(), progressTracker, sessions, true)
 
+    /**
+     * Notarise the given transaction and broadcast it to all the participants.
+     *
+     * @param transaction What to commit.
+     * @param sessions A collection of [FlowSession]s for each non-local participant of the transaction. Sessions to non-participants can
+     * also be provided.
+     * @param statesToRecord Which transactions to commit to the vault. Defaults to [StatesToRecord.ONLY_RELEVANT].
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt]: There's no default on this c'tor. Instead you probably want to mention the default value in the kdocs for the other c'tors.

Comment (rogersanick) [reply core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt]: Removed! Thanks, @shamsasari 

I've added a section in the initial comments explaining the impact of `StatesToRecord` on the initiating FinalityFlow. Is that sufficient? How can I indicate the default value in the kdocs for other c'tors (the @param annotation didn't seem appropriate given that it wasn't a param)

File: core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt [hunk 6]
```diff
@@ -76,6 +81,22 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
             progressTracker: ProgressTracker = tracker()
     ) : this(transaction, emptyList(), progressTracker, sessions, true)
 
+    /**
+     * Notarise the given transaction and broadcast it to all the participants.
+     *
+     * @param transaction What to commit.
+     * @param sessions A collection of [FlowSession]s for each non-local participant of the transaction. Sessions to non-participants can
+     * also be provided.
+     * @param statesToRecord Which transactions to commit to the vault.
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt]: Which _states_

Comment (rogersanick) [reply core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt]: Updated for both instances of the @statesToRecord param

**Review by shamsasari — APPROVED** [review]
